### PR TITLE
[WIP] [POC] Config based aliases

### PR DESCRIPTION
**Pull request by arturbosch**
TODO;
- rethink suppressing inside Context -> move and unify with rules report?
- MultiRule does actually double suppresion check
- remove aliases from issues
- refactor SuppressionSpec
- include into docs (generator)

### CODE REVIEW

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedImportsSpec.kt [hunk 1]
```diff
@@ -167,4 +167,17 @@ class UnusedImportsSpec : SubjectSpek<UnusedImports>({
 			assertThat(subject.lint(code)).isEmpty()
 		}
 	}
+
+	given("imports with aliases") {
+
+		it("should not report import as unused because the alias is used") {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedImportsSpec.kt]: This seems to be a relict from your previous PR.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Rule.kt [hunk 2]
```diff
@@ -18,8 +18,29 @@ abstract class Rule(override val ruleSetConfig: Config = Config.empty,
...
+	/**
+	 * A rule is motivated to point out a specific issue in the code base.
+	 */
 	abstract val issue: Issue
+
+	/**
+	 * An id this rule is identified with.
+	 * Conventionally the rule id is derived from the issue id as these two classes have a coexistence.
+	 */
 	final override val ruleId: String get() = issue.id
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Rule.kt]: Good comment!

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Rule.kt [hunk 3]
```diff
@@ -18,8 +18,29 @@ abstract class Rule(override val ruleSetConfig: Config = Config.empty,
...
 
-	override fun visitCondition(root: KtFile) = active && !root.isSuppressedBy(ruleId, issue.aliases)
+	/**
+	 * List of rule ids which can optionally be used in suppress annotations to refer to this rule.
+	 */
+	val aliases get() = valueOrDefault("aliases", emptySet<String>())
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Rule.kt]: Good comment!

**Comment by schalkms** [comment]
Perhaps we should make this also visible in detekt's documentation.

**Comment by arturbosch** [comment]
Wow, yeah maybe. Could be hard.
Well if the standard procedere is to override the alias defaults, it would just be a property we can identify by the generator. Will add this to the todo list.

**Comment by arturbosch** [comment]
The question is: ship this with RC10 to have it battle tested :D ?

**Pull request merged**